toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt

```diff
@@ -0,0 +1,95 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.inspections
+
+import com.intellij.codeInspection.*
+import com.intellij.psi.PsiElementVisitor
+import org.rust.cargo.CargoConstants
+import org.rust.ide.experiments.RsExperiments
+import org.rust.openapiext.isFeatureEnabled
+import org.rust.toml.crates.local.CratesLocalIndexService
+import org.rust.toml.isDependencyKey
+import org.toml.lang.psi.*
+import org.toml.lang.psi.ext.TomlLiteralKind
+import org.toml.lang.psi.ext.kind
+import org.toml.lang.psi.ext.name
+
+class CrateNotFoundInspection : LocalInspectionTool() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor =
+        buildVisitor(holder) ?: super.buildVisitor(holder, isOnTheFly)
+
+    private fun buildVisitor(holder: ProblemsHolder): PsiElementVisitor? {
+        if (!isFeatureEnabled(RsExperiments.CRATES_LOCAL_INDEX)) return null
+        if (holder.file.name != CargoConstants.MANIFEST_FILE) return null
+
+        return object : TomlVisitor() {
+            override fun visitKeyValue(element: TomlKeyValue) {
+                val table = element.parent as? TomlTable ?: return
+                val depTable = getDependencyTable(table) ?: return
+                if (depTable !is DependencyTableBase.DependencyTable) return
+
+                val segment = element.key.segments.getOrNull(0) ?: return
+                val name = segment.name ?: return
+                val value = element.value ?: return
+                if (value is TomlLiteral && value.kind is TomlLiteralKind.String) {
+                    handleDependency(Dependency(name, segment, mapOf("version" to value)), holder)
+                } else if (value is TomlInlineTable) {
+                    handleDependency(Dependency(name, segment, collectProperties(value)), holder)
+                }
+            }
+
+            override fun visitTable(element: TomlTable) {
+                val depTable = getDependencyTable(element) ?: return
+                if (depTable !is DependencyTableBase.SpecificDependencyTable) return
+
+                handleDependency(Dependency(depTable.crateName, depTable.crate, collectProperties(element)), holder)
+            }
+        }
+    }
+
+    private fun handleDependency(dependency: Dependency, holder: ProblemsHolder) {
+        if (CratesLocalIndexService.getInstance().getCrate(dependency.name) == null) {
+            holder.registerProblem(dependency.nameElement, "Crate ${dependency.name} not found")
+        }
+    }
+}
+
+data class Dependency(val name: String, val nameElement: TomlKeySegment, val properties: Map<String, TomlValue>)
```

Comment:
> Potential memory leak: don't hold PsiElement, use SmartPsiElementPointer instead 

Let's ultimately get rid of this class and pass its values separately. Also, `properties` seems to be unused

Reply:
I was already preparing a bit for other inspections, the Dependency code from this file will be probably refactored into a separate visitor to be used by multiple other inspections that might require the properties etc. But you are right that currently it's used only for this one inspection.